### ISSUE TITLE
RightCurly: Add support for ARRAY_INIT token type

### ISSUE THREAD

**Issue by eric-milles** [checkstyle/checkstyle#3325]
Similar problem with LeftCurly was moved to #3424 

$ cat ArrayInit.java

``` java
class ArrayInit
{
    String[] compact = {"value"}; // no violations -- want simple initializer to be compact

    String[] wantCompacted =
    {
        "value"}; // want simple initializer to be compact, but that may be a stretch for given config

    String[] wantAddOrReorderAsOneLineChange =
    {
        "value1",
        "value2",}; // VIOLATION -- want the closing brace of multi-line init to be on new line


    // more examples using the more verbose form of array init:

    String[] verboseButCompact = new String[] {"value"}; // no violations -- want simple initializer to be compact

    String[] verboseButWantCompacted = new String[]
    {
        "value"}; // want simple initializer to be compact, but that may be a stretch for given config

    String[] verboseButWantBracesInDifferentPositions = new String[]
    {
        "value1",
        "value2",}; // VIOLATION -- want the closing brace of multi-line init to be on new line
}
```

$ cat config.xml

``` xml
<!DOCTYPE module PUBLIC
  "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
  "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="RightCurly">
            <message key="line.alone" value="Closing ''{0}'' should be alone on a line" />
            <message key="line.new"   value="Closing ''{0}'' should be alone on a line" />
            <property name="option" value="alone" />
            <property name="tokens" value="CLASS_DEF,METHOD_DEF,CTOR_DEF,STATIC_INIT,INSTANCE_INIT,
                                           LITERAL_IF,LITERAL_ELSE,LITERAL_FOR,LITERAL_DO,LITERAL_WHILE,LITERAL_TRY,LITERAL_CATCH,LITERAL_FINALLY"
            />
        </module>
    </module>
</module>
```

$ java -jar checkstyle-7.0-all.jar -c config.xml ArrayInit.java
Starting audit...
Audit done.

---

Can ARRAY_INIT be added to the supported list of token types for RightCurly?  If supported, I could create a module like 

```
<module name="RightCurly">
  <property name="option" value="alone_or_singleline"/>
  <property name="tokens" value="ARRAY_INIT"/>
</module>
```

to allow compact array initializers for simple cases and well-formed multi-line initializers when wrapped by the formatter.  That is, when wrapped we want to see:

``` java
String[] strings = {
  "one",
  "two",
  "three",
};

and

Arrays.asList(new String[] {
  "one",
  "two",
  "three",
});
```

_I know Arrays.asList is variadic, but I wanted a familiar example to show passing an inline array._
##

**Comment by MEZk** [comment]
@romani 
Might be a problem as SAME brace policy says that the brace should be on the same line as the next part of a multi-block statement (one that directly contains multiple blocks: if/else-if/else or try/catch/finally). RightCurlyCheck now follows that rule.

**Comment by romani** [comment]
@MEZk ,
I do not see a problem to extend a allowed token set. It will give user ability to do smth valuable with formatting. If you know smth - please share an example.

**Comment by MEZk** [comment]
@romani 
Now the whole check is designed to follow the rule:
1) SAME: the brace should be on the same line **as the next part of a multi-block statement** (one that directly contains multiple blocks: **if/else-if/else or try/catch/finally**). Note, that the current design does not say that right curly should be on the same line as previous statement or any statement on the line.
2) ALONE: the brace should be alone on the line. Looks ok for arrays initialization braces.

If we add ARRAY_INIT into acceptable tokens set, it might lead to redesigning of the check logic.

**Comment by eric-milles** [comment]
I'm not sure I understand the "dubious" labeling.  Checkstyle exists to check and enforce style guidelines.  I clearly expressed the style we are expecting for multi-line array initializers, which require opening and closing braces.  There is a check for trailing comma.  Why can't the trailing brace also be positioned accordingly?

**Comment by romani** [comment]
@eric-milles ,
this label is for me to keep track that smth need be investigated more throughly. I approved this issue but MEZk , who was in that Check recently, mention that some problems might be ..... . Smb need to prove his concern or prove that there will be conflicts.

**Comment by romani** [comment]
this issue will be addressed by new Check - #5945
